### PR TITLE
Add ingress-operator in the openfaas Helm chart

### PR DESCRIPTION
**Pull request by jonatasbaldin**
Signed-off-by: Jonatas Baldin <jonatas.baldin@gmail.com>

## Description
Adding the `ingress-operator` to the Helm chart.

## TODO
This a list of potential things that I think still needs to be done. Please provide feedback here:
- [ ] Create an `amr64` image version of `ingress-operator` (to add in `values-arm64.yaml`)
- [ ] Bump the chart version

## Motivation and Context
Solves #502.
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?
I tested by upgrading the Helm release with new templates/values and by deploying this new version from scratch.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by jonatasbaldin** [comment]
Some mess with the commits because of forgetting to sign off :< But I fixed it and there's just the three I needed heh

**Comment by alexellis** [comment]
Thanks for working on the PR.

I'd like to see the IngressOperator only outputted into the armhf/arm64 YAML if a user has set it to `enabled: true`.

By default we need to set it to `enabled: false` everywhere, but have an override in the ARM files to make certain. We had an issue relating to this with ARM64 recently.

**Comment by jonatasbaldin** [comment]
Hey Alex! Just disabled it by default everywhere and also added the option in the armhr and arm64 (even tho the later has no docker images yet).

File: chart/openfaas/README.md [hunk 1]
```diff
@@ -283,6 +283,10 @@ Additional OpenFaaS options in `values.yaml`.
 | `operator.createCRD` | Create the CRD for OpenFaaS Function definition | `true` |
 | `ingress.enabled` | Create ingress resources | `false` |
 | `faasnetes.httpProbe` | Use a httpProbe instead of exec | `false` |
+| `ingressOperator.create` | Create the ingress-operator component | `true` |
```
Comment (alexellis) [inline chart/openfaas/README.md]: I thought we were going to go with `false` by default?

File: chart/openfaas/templates/ingress-operator-crd.yaml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+{{- if .Values.ingressOperator.create }}
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: functioningresses.openfaas.com
```
Comment (alexellis) [inline chart/openfaas/templates/ingress-operator-crd.yaml]: I haven't tested the new CRD with the operator's Function CRD yet.. could you confirm that both work on a new cluster before I merge?

File: chart/openfaas/values.yaml [hunk 3]
```diff
@@ -116,6 +116,16 @@ ingress:
   tls:
     # Secrets must be manually created in the namespace.
 
+ingressOperator:
```
Comment (alexellis) [inline chart/openfaas/values.yaml]: Let's add a comment?

```
# ingressOperator (optional) component to get TLS certs for Functions
#  - https://github.com/openfaas-incubator/ingress-operator
```

File: chart/openfaas/values.yaml [hunk 4]
```diff
@@ -116,6 +116,16 @@ ingress:
...
+  create: false
+  resources:
+    limits:
+      memory: 128Mi
+    requests:
+      memory: 25Mi
+
 # faas-idler configuration
 faasIdler:
   image: openfaas/faas-idler:0.2.1
```
Comment (alexellis) [inline chart/openfaas/values.yaml]: Relatedly I think faasIdler should have some limits too.

**Comment by alexellis** [comment]
After merge, a core team member will regenerate the charts and bump the version.

File: chart/openfaas/values.yaml (hunk 4 above)
Comment (alexellis) [reply chart/openfaas/values.yaml]: Ignore that comment.. it does have them.

File: chart/openfaas/values.yaml [hunk 5]
```diff
@@ -116,6 +116,16 @@ ingress:
   tls:
     # Secrets must be manually created in the namespace.
 
+ingressOperator:
+  image: openfaas/ingress-operator:0.4.0
+  replicas: 1
+  create: false
+  resources:
+    limits:
+      memory: 128Mi
```
Comment (alexellis) [inline chart/openfaas/values.yaml]: As a rule we don't set hard limits in the helm chart and leave that up to the user to decide. The requests values are set by us though.

File: chart/openfaas/README.md (hunk 1 above)
Comment (jonatasbaldin) [reply chart/openfaas/README.md]: Fixed.

File: chart/openfaas/values.yaml (hunk 3 above)
Comment (jonatasbaldin) [reply chart/openfaas/values.yaml]: Added.

File: chart/openfaas/values.yaml (hunk 5 above)
Comment (jonatasbaldin) [reply chart/openfaas/values.yaml]: Ok, fixed!

File: chart/openfaas/templates/ingress-operator-crd.yaml (hunk 2 above)
Comment (jonatasbaldin) [reply chart/openfaas/templates/ingress-operator-crd.yaml]: I didn't understand what I should test.

Is this it?
- Create a new cluster
- Create a function with the `Function` CRD
- Create a `FunctionIngress` for that function

If not, could you give me some more information?

Cheers!

Comment (alexellis) [reply chart/openfaas/templates/ingress-operator-crd.yaml]: Yep basically that please.

**Comment by alexellis** [comment]
I've tested that the two API resources / CRDs can be used together. I haven't tested the helm chart, but will merge given that the impact is low and @jonatasbaldin can re-test after that e2e.

Alex

**Comment by jonatasbaldin** [comment]
Hey there! Just tested the merged version and everything seems alright :shipit:

**Pull request merged**